pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -929,28 +983,42 @@ func (c *Reconciler) updatePipelineRunStatusFromInformer(ctx context.Context, pr
 		logger.Errorf("could not list TaskRuns %#v", err)
 		return err
 	}
-	pr.Status = updatePipelineRunStatusFromTaskRuns(logger, pr.Name, pr.Status, taskRuns)
+	updatePipelineRunStatusFromTaskRuns(logger, pr, taskRuns)
+
+	runs, err := c.runLister.Runs(pr.Namespace).List(labels.SelectorFromSet(pipelineRunLabels))
+	if err != nil {
+		logger.Errorf("could not list Runs %#v", err)
+		return err
+	}
+	updatePipelineRunStatusFromRuns(logger, pr, runs)
+
 	return nil
 }
 
-func updatePipelineRunStatusFromTaskRuns(logger *zap.SugaredLogger, prName string, prStatus v1beta1.PipelineRunStatus, trs []*v1beta1.TaskRun) v1beta1.PipelineRunStatus {
+func updatePipelineRunStatusFromTaskRuns(logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, trs []*v1beta1.TaskRun) {
 	// If no TaskRun was found, nothing to be done. We never remove taskruns from the status
 	if trs == nil || len(trs) == 0 {
-		return prStatus
+		return
 	}
 	// Store a list of Condition TaskRuns for each PipelineTask (by name)
 	conditionTaskRuns := make(map[string][]*v1beta1.TaskRun)
 	// Map PipelineTask names to TaskRun names that were already in the status
 	taskRunByPipelineTask := make(map[string]string)
-	if prStatus.TaskRuns != nil {
-		for taskRunName, pipelineRunTaskRunStatus := range prStatus.TaskRuns {
+	if pr.Status.TaskRuns != nil {
+		for taskRunName, pipelineRunTaskRunStatus := range pr.Status.TaskRuns {
 			taskRunByPipelineTask[pipelineRunTaskRunStatus.PipelineTaskName] = taskRunName
 		}
 	} else {
-		prStatus.TaskRuns = make(map[string]*v1beta1.PipelineRunTaskRunStatus)
+		pr.Status.TaskRuns = make(map[string]*v1beta1.PipelineRunTaskRunStatus)
 	}
 	// Loop over all the TaskRuns associated to Tasks
 	for _, taskrun := range trs {
+		// Only process TaskRuns that are owned by this PipelineRun.
+		// This skips TaskRuns that are indirectly created by the PipelineRun (e.g. by custom tasks).
+		if len(taskrun.OwnerReferences) < 1 || taskrun.OwnerReferences[0].UID != pr.ObjectMeta.UID {
```

Comment:
What is the purpose of this check here? Runs for custom tasks are created under runs not taskruns so they should not be part of this list here 🤔

Reply:
If a PipelineRun includes a Run that creates TaskRuns (such as the looping task), the PipelineRun's labels will be propagated to the TaskRuns created by the Run.  We don't want the PipelineRun adopting those TaskRuns.  It should only adopt the ones it directly created as part of the pipeline.